### DISCUSSION TITLE
Trouble getting mDNS discovery to work.

### DISCUSSION THREAD

**Discussion by barthelmeh** [libp2p/rust-libp2p#5524] [Q&A] [answered]
Hi there!

My university has a new course teaching p2p with libp2p and rust. We followed through the chat app example in a tutorial and I think that I have an understanding of how it (kinda) works; however, I'm unable to get my code to work.

I've copied the example exactly, including the .toml file, from the chat example, and have also used code that a friend wrote that works on his computer, but I am unable to get the mDNS discovery part working.
When I run the code using ```cargo run``` in two separate terminals, the application starts running and prints out the ```"Your node is listening on ..."``` messages but no mDNS connection is ever made between the terminals.

I am using a 2023 MacBook Pro running macOS Sonoma 14.1 and don't have much of an idea where to start. I have run the terminals with sudo permissions but still has the same issue. 
I have also run the following command: ```sudo launchctl list | grep -i mdnsresponder``` which gives me the following result:
```
520	0	com.apple.mDNSResponderHelper.reloaded
503	0	com.apple.mDNSResponder.reloaded
```

My understanding is that mDNS is working but I don't have much of a clue as to why the peers are unable to discover each other.

Any help is much appreciated!

**Comment by dariusc93** [comment, marked as answer]
Most of the time mdns doesnt work because either

A) The firewall on the machine is blocking it (or the service on the machine is disabled or need permission, which might be the case for mac); or
B) The router or modem used is blocking it

Could you provide the logs from when you ran the example using RUST_LOG=libp2p=debug ?

**Reply by barthelmeh** [reply]
Thanks for your reply.

I must say that I'm very new to rust in general so I may be doing this wrong; however, I ran the following:
```bash
export RUST_LOG=libp2p=debug
cargo run
```

My output was exactly the same as without the export command so I must be doing something wrong although I'm not quite sure. Please do tell me if I am doing something wrong as I am very new to rust.
I have looked through the firewall and even turned it off completely before running the program again with no success.

Although something strange did happen today, I didn't change anything and ran the program again and it was working! I restarted my computer to try and see if it would still work after that but after a restart I couldn't get it working again.

Thanks
